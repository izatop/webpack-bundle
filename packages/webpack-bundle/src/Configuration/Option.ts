export abstract class Option<T> {
    protected readonly value?: T;

    constructor(value?: T) {
        if (value) {
            this.value = value;
        }
    }

    public get name() {
        return this.constructor.name;
    }

    public getValue() {
        return this.value;
    }

    public abstract serialize(): unknown;
}
