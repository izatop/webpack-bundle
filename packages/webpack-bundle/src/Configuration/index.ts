export * from "./Options/AMD";
export * from "./Options/Bail";
export * from "./Options/Cache";
export * from "./Options/Context";
export * from "./Options/DevTool";
export * from "./Options/DevServer";
export * from "./Options/Entry";
export * from "./Options/Externals";
export * from "./Options/Mode";
export * from "./Options/Module";
export * from "./Options/ModuleRule";
export * from "./Options/Name";
export * from "./Options/Node";
export * from "./Options/Optimization";
export * from "./Options/Output";
export * from "./Options/Parallelism";
export * from "./Options/Performance";
export * from "./Options/Plugins";
export * from "./Options/Profile";
export * from "./Options/RecordsInputPath";
export * from "./Options/RecordsOutputPath";
export * from "./Options/RecordsPath";
export * from "./Options/Resolve";
export * from "./Options/ResolveLoader";
export * from "./Options/Stats";
export * from "./Options/Target";
export * from "./Options/Watch";
export * from "./Options/WatchOptions";
