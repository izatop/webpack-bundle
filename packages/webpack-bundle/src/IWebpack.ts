import {Configuration} from "webpack";

export type IWebpackKey = keyof Configuration;
export type IWebpackValue<K extends IWebpackKey = IWebpackKey> = Configuration[K];
