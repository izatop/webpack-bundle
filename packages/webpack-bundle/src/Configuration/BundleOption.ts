import * as webpack from "webpack";
import {IWebpackKey, IWebpackValue} from "../IWebpack";
import {Option} from "./Option";

export type OptionType = IWebpackValue<IWebpackKey>;

export class BundleOption<T extends OptionType = OptionType> extends Option<T> {
    public get key() {
        return this.name.replace(/^[A-Z]+/g, (chunk: string) => chunk.toLowerCase());
    }

    public serialize(): webpack.Configuration {
        return {
            [this.key]: this.value,
        };
    }
}
