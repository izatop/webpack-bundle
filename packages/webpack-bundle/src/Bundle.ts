import * as webpack from "webpack";
import {BundleOption} from "./Configuration/BundleOption";

export class Bundle {
    protected options: BundleOption[] = [];

    constructor(...option: BundleOption[]) {
        this.set(...option);
    }

    public set(...option: BundleOption[]) {
        this.options.push(...option);
    }

    public get<T extends BundleOption>(type: new() => T): T | undefined {
        return this.options.find((item) => item instanceof type) as T;
    }

    public getWebpackConfig() {
        const config: webpack.Configuration = {};
        for (const option of this.options) {
            Object.assign(config, option.serialize());
        }

        return config;
    }
}
