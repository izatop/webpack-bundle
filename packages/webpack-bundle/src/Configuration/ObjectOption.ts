import {IWebpackKey, IWebpackValue} from "../IWebpack";
import {BundleOption} from "./BundleOption";

export abstract class ObjectOption<T extends IWebpackValue<IWebpackKey>> extends BundleOption<T> {}
