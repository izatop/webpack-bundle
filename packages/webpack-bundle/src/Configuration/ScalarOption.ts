import {IWebpackKey, IWebpackValue} from "../IWebpack";
import {BundleOption} from "./BundleOption";

export abstract class ScalarOption<K extends IWebpackKey,
    T extends IWebpackValue<K> = IWebpackValue<K>> extends BundleOption<T> {
}
