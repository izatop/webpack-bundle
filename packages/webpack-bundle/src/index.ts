import {Bundle} from "./Bundle";
import * as Options from "./Configuration";
import {Loader} from "./Loader/Loader";

export {Bundle, Loader, Options};
export * from "./Configuration/Option";
export * from "./Configuration/ScalarOption";
export * from "./Configuration/ObjectOption";
export * from "./Configuration/BundleOption";
export * from "./IWebpack";
