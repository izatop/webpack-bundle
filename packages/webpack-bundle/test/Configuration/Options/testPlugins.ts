import * as webpack from "webpack";
import {Plugins} from "../../../src/Configuration";

test("", () => {
    const plugins = new Plugins([]);
    plugins.add(new webpack.EnvironmentPlugin({}));
    expect(plugins).toMatchSnapshot();

    plugins.replace(new webpack.EnvironmentPlugin({foo: 1}));
    expect(plugins).toMatchSnapshot();
});
