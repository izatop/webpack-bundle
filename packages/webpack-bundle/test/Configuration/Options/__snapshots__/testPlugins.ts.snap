// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Plugins {
  "value": Array [
    EnvironmentPlugin {
      "defaultValues": Object {},
      "keys": Array [],
    },
  ],
}
`;

exports[` 2`] = `
Plugins {
  "value": Array [
    EnvironmentPlugin {
      "defaultValues": Object {
        "foo": 1,
      },
      "keys": Array [
        "foo",
      ],
    },
  ],
}
`;
